File: python/code/binary_search.py
import csv
import time
import os
from tkinter import filedialog, messagebox
import tkinter as tk
from typing import List, Tuple


class Pair:
    def __init__(self, number: int, text: str):
        self.number = number
        self.text = text


def parse_file(filename: str) -> List[Pair]:
    """Parse CSV file and return list of Pair objects."""
    data_list = []
    try:
        with open(filename, 'r', newline='', encoding='utf-8') as file:
            reader = csv.reader(file)
            for row in reader:
                if len(row) >= 2:
                    number = int(row[0].strip())
                    text = row[1].strip()
                    data_list.append(Pair(number, text))
        return data_list
    except (IOError, ValueError) as e:
        print(f"Error reading file: {filename}")
        print(f"Error details: {e}")
        return None


def binary_search_single(data_list: List[Pair], target: int) -> int:
    """Perform binary search and return number of comparisons."""
    left = 0
    right = len(data_list) - 1
    comparisons = 0
    
    while left <= right:
        middle = (left + right) // 2
        comparisons += 1
        
        middle_value = data_list[middle].number
        if middle_value == target:
            return comparisons
        elif middle_value < target:
            left = middle + 1
        else:
            right = middle - 1
    
    return comparisons


def binary_search_analysis(data_list: List[Pair]) -> int:
    """Perform comprehensive binary search analysis."""
    output = []
    input_list_size = len(data_list)
    TIMING_ITERATIONS = 1000
    
    # ---------------------------- WARM UP -----------------------------
    print("Warming up...")
    for i in range(1000):
        binary_search_single(data_list, data_list[i % input_list_size].number)
    
    # ---------------------------- BEST CASE -----------------------------
    left = 0
    right = len(data_list) - 1
    best_case_target = data_list[(left + right) // 2].number
    
    best_case_start_time = time.perf_counter()
    best_case_comparisons = 0
    for i in range(TIMING_ITERATIONS):
        best_case_comparisons = binary_search_single(data_list, best_case_target)
    best_case_end_time = time.perf_counter()
    
    best_case_run_time = (best_case_end_time - best_case_start_time) * 1000 / TIMING_ITERATIONS
    output.append(f"Best Case Time   : {best_case_run_time:.6f} ms")
    
    # ---------------------------- WORST CASE -----------------------------
    worst_case_targets = []
    max_comparisons = 0
    
    # Find elements requiring maximum comparisons
    for i in range(input_list_size):
        comparisons = binary_search_single(data_list, data_list[i].number)
        if comparisons > max_comparisons:
            max_comparisons = comparisons
            worst_case_targets.clear()
            worst_case_targets.append(data_list[i].number)
        elif comparisons == max_comparisons:
            worst_case_targets.append(data_list[i].number)
    
    # Value greater than all elements
    target_greater = data_list[input_list_size - 1].number + 1
    comparisons_greater = binary_search_single(data_list, target_greater)
    if comparisons_greater >= max_comparisons:
        if comparisons_greater > max_comparisons:
            max_comparisons = comparisons_greater
            worst_case_targets.clear()
        worst_case_targets.append(target_greater)
    
    # Value smaller than all elements
    target_smaller = data_list[0].number - 1
    comparisons_smaller = binary_search_single(data_list, target_smaller)
    if comparisons_smaller >= max_comparisons:
        if comparisons_smaller > max_comparisons:
            max_comparisons = comparisons_smaller
            worst_case_targets.clear()
        worst_case_targets.append(target_smaller)
    
    worst_case_start_time = time.perf_counter()
    for i in range(TIMING_ITERATIONS):
        binary_search_single(data_list, worst_case_targets[i % len(worst_case_targets)])
    worst_case_end_time = time.perf_counter()
    
    worst_case_run_time = (worst_case_end_time - worst_case_start_time) * 1000 / TIMING_ITERATIONS
    output.append(f"Worst Case Time  : {worst_case_run_time:.6f} ms")
    
    # ---------------------------- AVERAGE CASE -----------------------------
    total_comparisons = 0
    total_search_time = 0
    
    for i in range(input_list_size):
        target = data_list[i].number
        
        start_time = time.perf_counter()
        comparisons = binary_search_single(data_list, target)
        end_time = time.perf_counter()
        
        total_search_time += (end_time - start_time)
        total_comparisons += comparisons
    
    total_time_ms = total_search_time * 1000
    average_time_ms = total_time_ms / input_list_size
    average_comparisons = total_comparisons / input_list_size
    
    output.append(f"Average Case Time: {average_time_ms:.6f} ms")
    
    # Write results to file
    write_to_file(f"python/output/binary_search/binary_search_{input_list_size}.txt", "\n".join(output))
    
    return total_comparisons


def write_to_file(path: str, content: str):
    """Write content to file, creating directories if needed."""
    try:
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, 'w', encoding='utf-8') as file:
            file.write(content)
        print(f"Output written to: {path}")
    except IOError as e:
        print("Failed to write output file.")
        print(f"Error: {e}")


def main():
    """Main function to handle file selection and run analysis."""
    # Create root window and hide it
    root = tk.Tk()
    root.withdraw()
    
    # Use file dialog to select CSV file
    file_path = filedialog.askopenfilename(
        title="Select a CSV dataset file for binary search",
        filetypes=[("CSV files", "*.csv"), ("All files", "*.*")]
    )
    
    if file_path:
        print(f"File FOUND: {file_path}")
        
        data_list = parse_file(file_path)
        if data_list is None:
            print("Failed to read or parse the file.")
            return
        
        if len(data_list) == 0:
            print("No data found in the file.")
            return
        
        binary_search_analysis(data_list)
    else:
        print("No file selected. Exiting.")
    
    root.destroy()


if __name__ == "__main__":
    main()

File: java/code/binary_search.java
import java.io.*;
import java.util.*;
import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;

public class binary_search {

    static class Pair {
        long number;
        String text;

        Pair(long number, String text) {
            this.number = number;
            this.text = text;
        }
    }

    public static void main(String[] args) {
        File file = null;

        // Use JFileChooser to select CSV file
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setDialogTitle("Select a CSV dataset file for binary search");
        fileChooser.setFileFilter(new FileNameExtensionFilter("CSV files", "csv"));
        int result = fileChooser.showOpenDialog(null);

        if (result == JFileChooser.APPROVE_OPTION) {
            file = fileChooser.getSelectedFile();
            String fullPath = file.getAbsolutePath();
            System.out.println("File FOUND: " + fullPath);

            List<Pair> dataList = ParseFile(fullPath);
            if (dataList == null) {
                System.out.println("Failed to read or parse the file.");
                return;
            }

            binarySearch(dataList);
        } else {
            System.out.println("No file selected. Exiting.");
        }
    }

    private static List<Pair> ParseFile(String filename) {
        List<Pair> list = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                long number = Long.parseLong(parts[0].trim());
                String text = parts[1].trim();
                list.add(new Pair(number, text));
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error reading file: " + filename);
            return null;
        }
        return list;
    }

    private static int binarySearch(List<Pair> list) {
        StringBuilder output = new StringBuilder();
        int inputListSize = list.size();
        final int TIMING_ITERATIONS = 1000;

        /// ---------------------------- WARM UP JVM -----------------------------
        System.out.println("Warming up JVM...");
        for (int i = 0; i < 1000; i++) {
            binarySearchSingle(list, list.get(i % inputListSize).number);
        }

        // ---------------------------- BEST CASE -----------------------------
        int left = 0;
        int right = list.size() - 1;
        long bestCaseTarget = list.get((left + right) / 2).number;

        long bestCaseStartTime = System.nanoTime();
        int bestCaseComparison = 0;
        for (int i = 0; i < TIMING_ITERATIONS; i++) {
            bestCaseComparison = binarySearchSingle(list, bestCaseTarget);
        }
        long bestCaseEndTime = System.nanoTime();

        double bestCaseRunTime = (bestCaseEndTime - bestCaseStartTime) / 1_000_000.0 / TIMING_ITERATIONS;
        output.append(String.format("Best Case Time   : %.6f ms\n", bestCaseRunTime));
        // output.append("Best Case Comparisons: " + bestCaseComparison + "\n\n");

        // ---------------------------- WORST CASE -----------------------------
        List<Long> worstCaseTargets = new ArrayList<>();
        int maxComparisons = 0;

        for (int i = 0; i < inputListSize; i++) {
            int comparisons = binarySearchSingle(list, list.get(i).number);
            if (comparisons > maxComparisons) {
                maxComparisons = comparisons;
                worstCaseTargets.clear(); // clear smaller comparison
                worstCaseTargets.add(list.get(i).number); // add the bigger comparison
            } else if (comparisons == maxComparisons) {
                worstCaseTargets.add(list.get(i).number);
            }
        }

        // Value greater than all elements
        long targetGreater = list.get(inputListSize - 1).number + 1;
        int comparisonsGreater = binarySearchSingle(list, targetGreater);
        if (comparisonsGreater >= maxComparisons) {
            if (comparisonsGreater > maxComparisons) {
                maxComparisons = comparisonsGreater;
                worstCaseTargets.clear();
            }
            worstCaseTargets.add(targetGreater);
        }

        // Value smaller than all elements
        long targetSmaller = list.get(0).number - 1;
        int comparisonsSmaller = binarySearchSingle(list, targetSmaller);
        if (comparisonsSmaller >= maxComparisons) {
            if (comparisonsSmaller > maxComparisons) {
                maxComparisons = comparisonsSmaller;
                worstCaseTargets.clear();
            }
            worstCaseTargets.add(targetSmaller);
        }
        /*
         * -----------------------TO DEBUG EACH ELEMENTS-------------------------------
         * for (int i = 0; i < inputListSize; i++) {
         * long target = list.get(i).number;
         * int comparisons = binarySearchSingle(list, target);
         * output.append(String.format("i: %d  comp: %d\n", i, comparisons));
         * 
         * }
         */

        long worstCaseStartTime = System.nanoTime();
        for (int i = 0; i < TIMING_ITERATIONS; i++) {
            binarySearchSingle(list, worstCaseTargets.get(i % worstCaseTargets.size()));
        }
        long worstCaseEndTime = System.nanoTime();

        double worstCaseRunTime = (worstCaseEndTime - worstCaseStartTime) / 1_000_000.0 / TIMING_ITERATIONS;
        output.append(String.format("Worst Case Time  : %.6f ms\n", worstCaseRunTime));
        // output.append("Worst Case Comparisons: " + maxComparisons + "\n");
        // output.append("Elements requiring max comparisons: " +
        // worstCaseTargets.size() + "\n\n");

        // ---------------------------- AVERAGE CASE -----------------------------
        long totalComparisons = 0;
        long totalSearchTime = 0;

        for (int i = 0; i < inputListSize; i++) {
            long target = list.get(i).number;

            long startTime = System.nanoTime();
            int comparisons = binarySearchSingle(list, target);
            long endTime = System.nanoTime();

            totalSearchTime += (endTime - startTime);
            totalComparisons += comparisons;
        }

        double totalTimeMs = totalSearchTime / 1_000_000.0;
        double averageTimeMs = totalTimeMs / inputListSize;
        double averageComparisons = (double) totalComparisons / inputListSize;

        // output.append(String.format("Total Average Case Time: %.6f ms\n",
        // totalTimeMs));
        output.append(String.format("Average Case Time: %.6f ms\n", averageTimeMs));
        // output.append(String.format("Average Comparisons per Search: %.6f\n",
        // averageComparisons));

        String baseDir = System.getProperty("user.dir"); // current working directory
        String outputPath = baseDir + File.separator + "java" + File.separator + "output" +
                File.separator + "binary_search" + File.separator +
                "binary_search_" + inputListSize + ".txt";

        writeToFile(outputPath, output.toString());

        return (int) totalComparisons;
    }

    private static int binarySearchSingle(List<Pair> list, long target) {
        int left = 0;
        int right = list.size() - 1;
        int comparisons = 0;

        while (left <= right) {
            int middle = (left + right) / 2;
            comparisons++;

            long middleValue = list.get(middle).number;
            if (middleValue == target) {
                return comparisons;
            } else if (middleValue < target) {
                left = middle + 1;
            } else {
                right = middle - 1;
            }
        }
        return comparisons;
    }

    private static void writeToFile(String path, String content) {
        try {
            File file = new File(path);
            file.getParentFile().mkdirs();
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
                writer.write(content);
            }
            System.out.println("Output written to: " + path);
        } catch (IOException e) {
            System.out.println("Failed to write output file.");
            e.printStackTrace();
        }
    }
}